sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -261,6 +262,78 @@ protected Symbol visitFunctionCall(FunctionCall node, ExpressionAnalysisContext
+                                          @Nullable Expression operandExpression,
+                                          @Nullable Expression defaultValue,
+                                          ExpressionAnalysisContext context) {
+            List<Symbol> operands = new ArrayList<>(whenClauseExpressions.size());
+            List<Symbol> results = new ArrayList<>(whenClauseExpressions.size());
+            Set<DataType> resultsTypes = new HashSet<>(whenClauseExpressions.size());
+
+            // check for global operand
+            Symbol operandLeftSymbol = null;
+            if (operandExpression != null) {
+                operandLeftSymbol = operandExpression.accept(innerAnalyzer, context);
+            }
+
+            for (WhenClause whenClause : whenClauseExpressions) {
+                Symbol operandRightSymbol = whenClause.getOperand().accept(innerAnalyzer, context);
+                Symbol operandSymbol = operandRightSymbol;
+
+                if (operandLeftSymbol != null) {
+                    FunctionIdent ident = new FunctionIdent(EqOperator.NAME,
+                        Arrays.asList(operandLeftSymbol.valueType(), operandRightSymbol.valueType()));
+                    operandSymbol = context.allocateFunction(
+                        new FunctionInfo(ident, EqOperator.RETURN_TYPE),
+                        Arrays.asList(operandLeftSymbol, operandRightSymbol)
+                    );
+                }
+
+                operands.add(operandSymbol);
+
+                Symbol resultSymbol = whenClause.getResult().accept(innerAnalyzer, context);
+                results.add(resultSymbol);
+                resultsTypes.add(resultSymbol.valueType());
+            }
+
+            if (resultsTypes.size() == 2 && !resultsTypes.contains(DataTypes.UNDEFINED)
+                || resultsTypes.size() > 2) {
+                throw new UnsupportedOperationException(
+                    "Data types of all result expressions of a CASE statement must be equal");
+            }
+
+            Symbol lastSymbol = null;
+            if (defaultValue != null) {
+                lastSymbol = defaultValue.accept(innerAnalyzer, context);
+            }
+
+            // process when clauses in reverse order
+            for (int i = operands.size() -1; i >= 0; i-- ) {
+                Symbol operand = operands.get(i);
+                Symbol result = results.get(i);
+                List<Symbol> arguments = Lists.newArrayList(operand, result);
+                List<DataType> argumentTypes = Lists.newArrayList(operand.valueType(), result.valueType());
+                if (lastSymbol != null) {
+                    arguments.add(lastSymbol);
+                    argumentTypes.add(lastSymbol.valueType());
+                }
+                lastSymbol = context.allocateFunction(
+                    getFunctionInfo(new FunctionIdent(IfFunction.NAME, argumentTypes)), arguments);
```

Comment:
same here, I think it would be great to move some of the logic into IfFunction. Or maybe just extract a function to process the operands. This createCaseFunction method is already rather large